src/Clapi/Relay.hs

```diff
@@ -93,44 +97,74 @@ toFrcud :: Namespace -> ProtoFrcUpdateDigest -> FrcUpdateDigest
+  Map.lookup ns vsm
+
+rLookupDef
+  :: (VsLookupDef def, MkSubErrIdx (Tagged def Seg))
+  => Map Namespace Valuespace -> Namespace -> Tagged def Seg
+  -> Either (SubErrorIndex, String) def
+rLookupDef vsm ns s =
+    vsmLookupVs ns vsm >>= first (mkSubErrIdx ns s,) . vsLookupDef s
+
+rGet
+  :: Map Namespace Valuespace -> Namespace -> Path
+  -> Either
+       (SubErrorIndex, String)
+       (Definition, Tagged Definition Seg, Editable, RoseTreeNode [WireValue])
+rGet vsm ns p =
+    vsmLookupVs ns vsm >>= first (mkSubErrIdx ns p,) . valuespaceGet p
+
 genInitDigests
-  :: ClientGetDigest -> Valuespace
+  :: ClientGetDigest -> Map Namespace Valuespace
   -> (OutboundClientSubErrsDigest, [OutboundClientInitialisationDigest])
-genInitDigests (ClientGetDigest ptGets tGets dGets) vs =
+genInitDigests (ClientRegs ptGets tGets dGets) vsm =
   let
-    (ptSubErrs, postDefs) = mapPartitionEither
-      $ Map.fromSet (flip vsLookupPostDef vs) ptGets
-    (tSubErrs, defs) = mapPartitionEither
-      $ Map.fromSet (flip vsLookupDef vs) tGets
-    (dSubErrs, treeData) = mapPartitionEither
-      $ Map.fromSet (flip valuespaceGet vs) dGets
+    g :: (a -> b -> Either c d) -> (a, b) -> Either c ((a, b), d)
+    g f x@(a, b) = second (x,) $ f a b
+
+    h
+      :: (Ord b, MkSubErrIdx b)
+      => (Namespace -> b -> Either (SubErrorIndex, String) d)
+      -> Mos Namespace b
+      -> (Map SubErrorIndex String, Map (Namespace, b) d)
+    h f m = bimap Map.fromList Map.fromList
+      $ partitionEithers $ g f <$> Mos.toList m
+
+    ptSubErrs, tSubErrs, dSubErrs :: Map SubErrorIndex String
+    postDefs :: Map (Namespace, Tagged PostDefinition Seg) PostDefinition
+    defs :: Map (Namespace, Tagged Definition Seg) Definition
+    treeData :: Map (Namespace, Path) _
+    (ptSubErrs, postDefs) = h (rLookupDef vsm) ptGets
+    (tSubErrs, defs) = h (rLookupDef vsm) tGets
+    (dSubErrs, treeData) = h (rGet vsm) dGets
 
     -- FIXME: can we only ever have a single error per SubErrIdx now(?)
     -- therefore could ditch the `pure`?
-    ocsed = fmap (pure . Text.pack)
-        $ Map.mapKeys PostTypeSubError ptSubErrs
-       <> Map.mapKeys TypeSubError tSubErrs
-       <> Map.mapKeys PathSubError dSubErrs
+    ocsed = fmap (pure . Text.pack) $ ptSubErrs <> tSubErrs <> dSubErrs
 
     ocids = Map.elems $ Map.mapWithKey toFrcud $ Map.unionsWith (<>)
-      [ translateDefData setPostDefs postDefs
-      , translateDefData setDefs defs
-      , translateTreeData treeData
+      [ i (toDefPfrcud setPostDefs) postDefs
+      , i (toDefPfrcud setDefs) defs
+      , i toDatPfrcud treeData
       ]
   in
     (ocsed, ocids)
   where
-    translateTreeData =
-        fmap fold
-      . snd
-      . nestMapsByKey splitNs
-      . Map.mapMaybeWithKey (\path x -> (mkPfrcud x) . snd <$> splitNs path)
-    mkPfrcud (def, ttn, ed, treeNode) path =
-      -- We could assert here that the namespace of the (Tagged TypeName) == ns
+    i
```

Comment:
This might want a name?

Reply:
It was hard to name the intermediate transformations better than their type signatures already explained what they did. After offline discussion, I think we'll leave these be for now. They might change when we bring the NamespaceTracker and the Relay together.